pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
-func (h *VMNetworkController) SetDefaultNetworkMacAddress(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
-	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil {
+func (h *VMNetworkController) SyncMacAddressAndControllerRevision(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil || vmi.Status.Phase != kubevirtv1.Running || len(vmi.Status.Interfaces) == 0 {
 		return vmi, nil
 	}
 
-	if vmi.Status.Phase != kubevirtv1.Running {
-		return vmi, nil
+	// ensure vmi mac is backfilled to vm
+	retryvmi := false
+	if vmi, err := h.syncNetworkDynamicMacAddress(vmi, &retryvmi); err != nil {
+		return vmi, err
+	}
+
+	// ensure vmi mac is backfilled to rc
+	retrycr := false
+	if vmi, err := h.syncControllerRevision(vmi, &retrycr); err != nil {
+		return vmi, err
 	}
 
-	if len(vmi.Status.Interfaces) == 0 {
+	// vmi may end onChange, but vm/rc is not synced yet
+	if retryvmi || retrycr {
+		h.vmiController.EnqueueAfter(vmi.Namespace, vmi.Name, time.Second*1)
 		return vmi, nil
 	}
 
-	err := h.updateVMDefaultNetworkMacAddress(vmi)
+	// last step: if vm still did not sync with rc, update it
+	return h.updateVM(vmi)
+}
+
+// if user does not specify mac address when creating a VM
+// guest OS will use the initial allocated mac address from its DHCP config
+// when VM restarts that mac address may change, and related IP is not persistent
+// syncNetworkDynamicMacAddress saves the dynamic mac address observed from the the VMI status to vm
+func (h *VMNetworkController) syncNetworkDynamicMacAddress(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
 	if err != nil {
 		return vmi, err
 	}
 
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	vmCopy := vm.DeepCopy()
+	vmiInterfaces := getVmiInterfaceMacs(vmi)
+	updated := false
+	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmIface.Name == "" || vmIface.MacAddress != "" {
+			continue
+		}
+		if macAddress := vmiInterfaces[vmIface.Name]; macAddress != "" {
+			// only backfill the network mac address when it is empty
+			logrus.Infof("set VM %s network %s macAddress to %s", vm.Name, vmIface.Name, macAddress)
+			vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces[i].MacAddress = macAddress
+			updated = true
+		}
+	}
+	if !updated {
+		return vmi, nil
+	}
+	*retry = true
+	if _, err := h.vmClient.Update(vmCopy); err != nil {
+		return vmi, fmt.Errorf("failed to update vmi mac to vm: %w", err)
+	}
+
 	return vmi, nil
 }
 
-func (h *VMNetworkController) updateVMDefaultNetworkMacAddress(vmi *kubevirtv1.VirtualMachineInstance) error {
-	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
-	if err != nil {
-		return err
+func getVMInterfaceMacs(vm *kubevirtv1.VirtualMachine) map[string]string {
```

Comment:
we could refactor this into a common method which has input of `[]devices.Interfaces` and return the map[string]string.
This way it can be used across both VM's and VMI's